lib/srv/app/listener.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"context"
+	"net"
+
+	"github.com/gravitational/trace"
+)
+
+// listener wraps a net.Conn in a net.Listener interface. This allows passing
+// a channel connection from the reverse tunnel subsystem to an HTTP server.
+type listener struct {
```

Comment:
This wrapper requires you to create a `listener` per incoming connection, right?
https://github.com/gravitational/teleport/blob/79e595a1019ed47ab1d8c4d1e1ee354d3c583d80/lib/reversetunnel/agentpool.go#L318-L361 (in my pending k8s PR) makes a `reversetunnel.ConnHandler` that is also a `net.Listener`, so you can reuse the same listener for you app server forever and only call `http.Server.Serve` once.
Your server then just needs to accept a `net.Listener` in params.

Reply:
Sure, I can switch to this once it's merged in.